File: day4.pl
#!perl
use v5.24;
use warnings;
use Path::Tiny qw(path);

my $valid = 0;

LINE: for my $line (path('input/day4.txt')->lines) {
  my %ok;
  for my $w (split /\s+/, $line) {
    my $k = join '', sort split //, $w;
    next LINE if $ok{$k};
    $ok{$k}++;
  }
  $valid++;
}

say "total valid: $valid";

File: 2017/day4.py
#!/usr/bin/env python

def valid_pw(pw):
    s = set()
    for w in pw.split():
        # k = w                  # easy case
        k = ''.join(sorted(w)) # hard case
        if k in s:
            return False
        s.add(k)
    return True

with open('input/day4.txt', 'r') as f:
    total = 0
    for line in f:
        if valid_pw(line):
            total += 1

print "total: %s" % total

